### ISSUE TITLE
Modifying list of items in-place causes items to be loaded multiple times

### ISSUE THREAD

**Issue by smcgivern** [exAspArk/batch-loader#44]
Consider this script:

```ruby
def modify!(xs)
  xs.map! { |x| x.map(&:succ) }
end

def load_value(x)
  BatchLoader.for(x).batch do |xs, loader|
    p xs
    modify!(xs).each { |x| loader.call(x.map(&:prev), x) }
  end
end

a = [1, 2, 3]
b = [4, 5, 6]

load_value(a)
load_value(b)
load_value(a)
```

I would expect the last `load_value` call here to not print anything. It actually prints `[[1, 2, 3], [4, 5, 6]]`.

We discovered this in GitLab because we were doing something similar; see https://gitlab.com/gitlab-org/gitlab-ce/issues/60829#note_163484873 for a fuller explanation. We shouldn't be doing that (and we'll fix it) but it might make sense for BatchLoader to be more robust here, too?

@exAspArk what do you think?

cc @DouweM

**Comment by exAspArk** [comment]
Hey @smcgivern, 

Thank you for opening this issue! I totally agree that `BatchLoader` should be more robust in this case. I'll look into it.

**Comment by smcgivern** [comment]
This is not stale! 😉

**Comment by exAspArk** [comment]
I released a new version `1.4.1` which freezes a list of items and doesn't allow mutating it. 

@smcgivern thank you!
